### PR TITLE
[HCAL] Switch to linear LUTs for 2018 TPs

### PR DESCRIPTION
**Pull request by matz-e**
Switch the output energy of HCAL TPs to a homogeneous baseline of 1/2 GeV LSB. The trigger-internal LSB is changed to 1/8 or 1/16 GeV LSB, depending on QIE-type of the inputs, to maximize dynamic range and accuracy.

This PR will need a new GT with metadata conditions to adjust energy thresholds, and new L1 calibrations once new RelVal samples are generated.

### CODE REVIEW

**Comment by nsmith-** [comment]
@matz-e will we need to alter the GTs in autoconditions for this to pass tests?

**Comment by matz-e** [comment]
No, tests should pass without any GT changes, those are just for thresholds.

**Comment by nsmith-** [comment]
please test

**Comment by abdoulline** [comment]
@matz-e  Matthias, it seems there is some conflict with what has been added to Eras in PR #21651, could you, please, take a look?

**Comment by arunhep** [comment]
@matz-e here is the commit with the GT changes : https://github.com/cms-sw/cmssw/commit/1ed6a4db5827d9eeb3c00e3232a68c97ef61607a

**Comment by matz-e** [comment]
@arunhep I don't seem to be able to pull that commit in my local area, and it does not seem to be associated with a branch?

**Comment by arunhep** [comment]
@matz-e this is the branch : arunhep:ak_HCALLUTs_2018
and this is the commit : https://github.com/arunhep/cmssw/commit/1ed6a4db5827d9eeb3c00e3232a68c97ef61607a

**Comment by matz-e** [comment]
@arunhep Thanks! I was confused as github didn't show you user/branching info in the link to the commit.

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

**Comment by ghellwig** [comment]
please test

**Comment by matz-e** [comment]
Sorry, there was one outstanding item that was just addressed in the HCAL Trigger DPG meeting.

**Comment by ghellwig** [comment]
@matz-e no worries!
Can you confirm that it is now ready for review?

**Comment by matz-e** [comment]
Yes, the PR is ready for review now! The last commit just took care of implementing the right thresholds for 2019 and 2023, which we hadn't looked at before.

**Comment by ghellwig** [comment]
please test

File: CalibCalorimetry/CaloTPG/plugins/CaloTPGTranscoderULUTs.cc [hunk 1]
```diff
@@ -60,10 +60,13 @@ class CaloTPGTranscoderULUTs : public edm::ESProducer {
   std::vector<int> ietah;
   std::vector<int> ZS;
   std::vector<int> LUTfactor;
+  bool linearLUTs_;
   double nominal_gain;
   double RCTLSB;
   int NCTScaleShift;
   int RCTScaleShift;  
+  double lsbQIE8;
+  double lsbQIE11;
```
Comment (ghellwig) [inline CalibCalorimetry/CaloTPG/plugins/CaloTPGTranscoderULUTs.cc]: configuration parameters should be declared const (-> they have to be read in the initializer list of the constructor)
this applies also to the existing parameters

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc [hunk 2]
```diff
@@ -35,14 +36,18 @@ void CaloTPGTranscoderULUT::loadHCALCompress(HcalLutMetadata const& lutMetadata,
...
 
     // Compute compression LUT
     for (unsigned int i=0; i < OUTPUT_LUT_SIZE; i++) {
 	analyticalLUT[i] = min((unsigned int)(sqrt(14.94*log(1.+i/14.94)*i) + 0.5), TPGMAX - 1);
-	linearRctLUT[i] = min((unsigned int)(i/rct_factor_), TPGMAX - 1);
-	linearNctLUT[i] = min((unsigned int)(i/nct_factor_), TPGMAX - 1);
+	linearQIE8LUT[i] = min((unsigned int)((i+.5)/lin8_factor_), TPGMAX - 1);
+	linearQIE11LUT[i] = min((unsigned int)((i+.5)/lin11_factor_), TPGMAX - 1);
+	linearRctLUT[i] = min((unsigned int)((i+.5)/rct_factor_), TPGMAX - 1);
+	linearNctLUT[i] = min((unsigned int)((i+.5)/nct_factor_), TPGMAX - 1);
```
Comment (ghellwig) [inline CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc]: please replace the C-style cast with `static_cast<unsigned int>`

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc [hunk 3]
```diff
@@ -89,18 +98,29 @@ void CaloTPGTranscoderULUT::loadHCALCompress(HcalLutMetadata const& lutMetadata,
...
-		   hcaluncomp_[index][tpg] = (tpg == 0 ? low : factor * mid);
-		   low = i;
-		   tpg = outputLUT_[index][i];
-		}
-	    }
-	    hcaluncomp_[index][tpg] = factor * low;
+           if (allLinear_) {
+              LUT tpg = outputLUT_[index][0];
+              hcaluncomp_[index][tpg] = 0;
+              for (unsigned int i = 0; i < getOutputLUTSize(id); ++i){
```
Comment (ghellwig) [inline CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc]: if you store `getOutputLUTSize(id)` in a variable, you do not have to compute it in each loop iteration

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc [hunk 4]
```diff
@@ -89,18 +98,29 @@ void CaloTPGTranscoderULUT::loadHCALCompress(HcalLutMetadata const& lutMetadata,
...
+                 if (outputLUT_[index][i] != tpg){
+                    tpg = outputLUT_[index][i];
+                    hcaluncomp_[index][tpg] = lsb_factor_ * i / (isOnlyQIE11(id) ? lin11_factor_ : lin8_factor_);
+                 }
+              }
+           } else {
+              double factor = nominal_gain_ / cosh_ieta * granularity;
+              LUT tpg = outputLUT_[index][0];
+              int low = 0;
+              for (unsigned int i = 0; i < getOutputLUTSize(id); ++i){
```
Comment (ghellwig) [inline CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc]: see above

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h [hunk 5]
```diff
@@ -34,7 +34,7 @@ class CaloTPGTranscoderULUT : public CaloTPGTranscoder {
   double hcaletValue(const HcalTrigTowerDetId& hid, const HcalTriggerPrimitiveSample& hc) const override;
   virtual bool HTvalid(const int ieta, const int iphi, const int version) const;
   virtual const std::vector<unsigned int> getCompressionLUT(const HcalTrigTowerDetId& id) const;
-  virtual void setup(HcalLutMetadata const&, HcalTrigTowerGeometry const&, int, int);
+  virtual void setup(HcalLutMetadata const&, HcalTrigTowerGeometry const&, int, int, double, double, bool);
```
Comment (ghellwig) [inline CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h]: for basic types like `int`, `double` and `bool` it is usually helpful to provide a parameter name, while this is, of course, not needed for expressive type names like `HcalLutMetadata`

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h [hunk 6]
```diff
@@ -62,13 +62,17 @@ class CaloTPGTranscoderULUT : public CaloTPGTranscoder {
...
   void loadHCALCompress(HcalLutMetadata const&, HcalTrigTowerGeometry const&) ; //Analytical compression tables
 
   // Member Variables
+  bool allLinear_ = false;
   double nominal_gain_;
   double lsb_factor_;
   double rct_factor_;
   double nct_factor_;
+  double lin8_factor_;
+  double lin11_factor_;
```
Comment (ghellwig) [inline CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h]: see my comment above concerning the configuration parameters

File: SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py [hunk 7]
```diff
@@ -4,13 +4,17 @@
...
+
 from SimCalorimetry.HcalTrigPrimProducers.hcaltpdigi_cfi import *
 from CalibCalorimetry.CaloTPG.CaloTPGTranscoder_cfi import *
 HcalTPGCoderULUT = cms.ESProducer("HcalTPGCoderULUT",
     read_Ascii_LUTs = cms.bool(False),
     read_XML_LUTs = cms.bool(False),
     read_FG_LUTs = cms.bool(False),
     LUTGenerationMode = cms.bool(True),
+    linearLUTs = cms.bool(False),
+    TPScales = TPScales,
```
Comment (ghellwig) [inline SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: CMS naming convention would suggest names starting with lower case, I think.
This would have to be propagated to the configs, of course.

**Review by ghellwig — CHANGES_REQUESTED** [review]

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

File: SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py (hunk 7 above)
Comment (matz-e) [reply SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: Whenever I try to look for a consistent CMS style, I get lost… is there something that a mere mortal can find within 60s of using Google? The FAQ on GH seem to be more concerned with CVS than with coding style!

I can rename *that* parameter (and more, if requested)… at some point this is going to become a gigantic code-reformatting exercise, though.

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc (hunk 3 above)
Comment (matz-e) [reply CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc]: Sorry, I overlooked that that is already a variable…

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h (hunk 5 above)
Comment (matz-e) [reply CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h]: Sure!

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h (hunk 6 above)
Comment (matz-e) [reply CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h]: **That** parameter is set by the `setup(…)` function though. How do I handle that better?

**Comment by arunhep** [comment]
@matz-e you can remove the autoCond update from this PR. This PR https://github.com/cms-sw/cmssw/pull/21689 (merged today) takes care of the updates needed for this PR.

**Comment by matz-e** [comment]
@arunhep Thanks, I've already done so! I'm just waiting for my checks to come back after addressing most of @ghellwig's comments.

**Comment by franzoni** [comment]
hello, has the L1T  DPG signed off on the re-calibration work which will need to follow ?
Is the choice of switching to linear LUTs for HCAL in 2018 operations from the start a made decision ? 

@rekovic @abrinke1

**Comment by abrinke1** [comment]
Hi @franzoni 
We plan to have new Layer-1 calibrations for next year no matter what.  I believe that we have signed off on the choice of switching to linear scale, but I will let Alex Zabi comment if he thinks there are still open questions.

**Comment by abdoulline** [comment]
@matz-e  Matthias, you're at the updates, right?  There might be still a hope to have it pre3.

**Comment by lpernie** [comment]
Since PR #21689  has been merged, autocond: should be updated looking at [1], i.e. using the "_v3" version of the 3 GT changed. @arunhep can further comment if needed.
[1] https://github.com/cms-sw/cmssw/blob/master/Configuration/AlCa/python/autoCond.py

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h (hunk 6 above)
Comment (ghellwig) [reply CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h]: Ok, looks like `setup` needs the EventSetup, i.e. you cannot know the value of these parameters at construction time.

**Comment by davidlange6** [comment]
please test

File: SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py (hunk 7 above)
Comment (ghellwig) [reply SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: I am also not aware of anything else than this:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideDevelopersGuide#Code_conventions
But as it says, the linked document is a bit outdated.
I actually make these comments on the convention when I saw such comments by others with a longer history in CMS code development, e.g. @slava77 makes a lot of such comments and might know a more up-to-date reference.

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h (hunk 6 above)
Comment (matz-e) [reply CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.h]: I just tried to remove the `setup` function in its entirety and move the logic to the constructor. There's another piece of code that depends on this class from L1 that in theory should be broken anyways? If we can remove that, it'll make the code a bit cleaner.

**Comment by arunhep** [comment]
@lpernie 
i made the comment already : https://github.com/cms-sw/cmssw/pull/21657#issuecomment-352462036

File: SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py (hunk 7 above)
Comment (matz-e) [reply SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: I think I read through those a few years back, when Jim Brooke pointed them out to me, and they are no more than common sense to me, and kept pretty vague, still. IMO it is at least missing a clarification of using camel case vs underscores (in variables), and that variable names should not look like classes, i.e., should start with a lowercase letter. I fixed one such case in the last commit.

A major omission is the lack of a whitespace definition: indentation, parenthesis placement, whitespace around operators, no one-line if/switch/case statements… I parse code partially by its visual structure, and having a complete mix of indentation styles is pretty annoying. `mkedanlzr` spat out a file with 3 space indent when I used it first, and since no-one seemed to agree on anything, that's what I use as a standard for CMSSW. I'm amazed that no-one is bothered enough by the constant re-formatting to actually enforce something homogeneous!\</rant>

Comment (ghellwig) [reply SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: there's a syntax error: you are missing the opening `<rant>` tag ;)

Comment (matz-e) [reply SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: Customary syntax error… :)

Anyhow, do you want me to change that `TPScales` to `tpScales`?

Comment (matz-e) [reply SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: FWIW, I would like to do so in a follow-up PR. I did a good axe-grind and removed a bunch of HCAL LUT-generation code that should never be able to run anyways, this would be a candidate for that, too.

Comment (ghellwig) [reply SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: I would prefer that you change this name here as you are introducing the parameter here.
Other parameters where this applies could be handled in a separate PR.

Comment (matz-e) [reply SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cff.py]: OK

**Comment by perrotta** [comment]
please test

File: RecoTracker/DeDx/plugins/DeDxEstimatorProducer.cc [hunk 8]
```diff
@@ -34,8 +34,8 @@ void DeDxEstimatorProducer::fillDescriptions(edm::ConfigurationDescriptions& des
   desc.add<edm::InputTag>("tracks",edm::InputTag("generalTracks"));
   desc.add<bool>("UsePixel",false); 
   desc.add<bool>("UseStrip",true); 
-  desc.add<double>("MeVperADCPixel",3.61e-06*265);
-  desc.add<double>("MeVperADCStrip",3.61e-06);
+  desc.add<double>("MeVperADCPixel",3.61e-06);
+  desc.add<double>("MeVperADCStrip",3.61e-06*265);
```
Comment (slava77) [inline RecoTracker/DeDx/plugins/DeDxEstimatorProducer.cc]: this change does not belong in this PR.
Please rebase appropriately to get rid of this change.

**Comment by deguio** [comment]
@slava77 
I think @matz-e is not around at the moment. I can do the rebase and open a new PR. would this be acceptable to have this development considered for pre3?
thanks,
F.

**Comment by slava77** [comment]
On 12/20/17 1:12 PM, Federico De Guio wrote:
> @slava77 <https://github.com/slava77>
> I think @matz-e <https://github.com/matz-e> is not around at the moment.
> I can do the rebase and open a new PR.


> would this be acceptable to have
> this development considered for pre3?

if it is done, no reco signature will be necessary.

**Comment by deguio** [comment]
cross reference.
rebase at #21756

**Comment by slava77** [comment]
-1

reco signature should not be necessary
#21756
